torchvision/csrc/io/image/cpu/decode_gif.cpp

```diff
@@ -0,0 +1,136 @@
+#include "decode_gif.h"
+#include "giflib/gif_lib.h"
+
+namespace vision {
+namespace image {
+
+typedef struct reader_helper_t {
+  torch::Tensor const* encoded_bytes;
+  int num_bytes_read; // number of bytes read so far in the tensor
+} reader_helper_t;
+
+// That function is used by GIFLIB routines to read the encoded bytes.
+// This reads `len` bytes and writes them into `buf`. The data is read from the
+// input tensor passed to decode_gif().
+int read_from_tensor(GifFileType* gifFile, GifByteType* buf, int len) {
+  // the UserData field was set in DGifOpen()
+  reader_helper_t* reader_helper = (reader_helper_t*)gifFile->UserData;
+  auto num_encoded_bytes = (int)reader_helper->encoded_bytes->numel();
+  auto data_ptr = reader_helper->encoded_bytes->data_ptr<GifByteType>();
+
+  auto i = 0;
+  auto num_bytes_to_read =
+      std::min(len, num_encoded_bytes - reader_helper->num_bytes_read);
+  while (i < num_bytes_to_read) {
+    buf[i] = data_ptr[reader_helper->num_bytes_read + i];
+    i++;
+  }
+  reader_helper->num_bytes_read += i;
```

Comment:
actually, why do you need to count this? Do you enter this function multiple times or from different threads? Can `reader_helper->num_bytes_read` be something else than 0?

Reply:
> Do you enter this function multiple times?
> Can reader_helper->num_bytes_read be something else than 0?

Yes and yes, the function is entered everytime the `InternalRead()` function of GIFLIB is called (`git grep` for it). It's basically called to read the headers and then to read the image data. `num_bytes_read` simply points to the current byte that was last read within the tensor . It's similar to how files are read in C with `fread`:

`fread(..., size=10)` reads the first 10 bytes, then calling `fread(..., size=5)` would read the next 5 bytes etc.